### PR TITLE
Add user bios

### PR DESCRIPTION
**Pull request by Axighi**
Added vscode launch config file.

### CODE REVIEW

**Comment by dessalines** [comment]
Thanks! Just going through this now.

File: server/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,102 @@
+{
```
Comment (dessalines) [inline server/.vscode/launch.json:1]: Could you remove this file, none of the main devs use vscode, and I don't really want any IDE specific stuff in the codebase.

File: ui/src/components/user.tsx [hunk 2]
```diff
@@ -570,6 +572,21 @@ export class User extends Component<any, UserState> {
...
+                  {i18n.t('bio')}
+                </label>
+                <div class="col-lg-9">
+                  <input
+                    id="user-bio"
+                    class="form-control"
+                    placeholder={i18n.t('optional')}
+                    value={this.state.userSettingsForm.bio}
+                    onInput={linkEvent(this, this.handleUserSettingsBioChange)}
+                    minLength={3}
```
Comment (dessalines) [inline ui/src/components/user.tsx:586]: Could you add a maxLength here? I'm not sure what a good bio max length should be.... 

If these are twitter style bios, (155 chars) I could see them being useful as a potential user-hover style thing.... But if they're really long "user pages", then I would only want them to be shown on the details page only.

File: server/src/api/user.rs [hunk 3]
```diff
@@ -557,6 +558,11 @@ impl Perform for Oper<SaveUserSettings> {
       None => read_user.email,
     };
 
+    let bio = match &data.bio {
+      Some(bio) => Some(bio.to_owned()),
+      None => read_user.bio,
+    };
```
Comment (dessalines) [inline server/src/api/user.rs:564]: This back-end will also need to check for too many characters in a bio, and throw an APIError if it is. That error string will have to be added to `en.json`.

**Comment by Nutomic** [comment]
Looks good and seems to work fine. But it should really use a multi-line input. Thats gonna take more space, so its a good reason to put the whole user settings on a separate page, not in the sidebar (but that can be done in a separate PR).

File: server/.vscode/launch.json (hunk 1 above)
Comment (Nutomic) [reply server/.vscode/launch.json:1]: Best just add it to .gitignore instead.

Comment (Axighi) [reply server/.vscode/launch.json:1]: Good idea, will do.

File: ui/src/components/user.tsx (hunk 2 above)
Comment (Axighi) [reply ui/src/components/user.tsx:586]: Where to put this maxLength constant?

**Comment by Axighi** [comment]
> Looks good and seems to work fine. But it should really use a multi-line input. Thats gonna take more space, so its a good reason to put the whole "edit user settings" on a separate page, not in the sidebar (but that can be done in a separate PR).

Agree. It should be done in another PR.

**Comment by Axighi** [comment]
> Looks pretty good, of course aside from bios not actually being shown to other users.
> This would require adding bio to the `user_view` and `user_fast_view` (and possibly comment and post views too if we want to show some cool tooltips there eventually).

I thought you wanted bio only shown here:
Screen Shot 2020-07-30 at 22 00 00
And this can be seen by other users. Where else would you like it to be displayed?

**Comment by dessalines** [comment]
Oh my bad lol, I didn't notice I already had bios in the `user_view`, I'd thought this needed to be added for some reason.

That's probably good for now.

But also ya I agree with nutomic its probably better to have that bio as a markdown-textarea, rather than a simple line input. Use it like 

```
              <MarkdownTextArea
                initialContent={this.state.siteForm.description}
                onContentChange={this.handleSiteDescriptionChange}
              />
```

File: ui/src/components/user.tsx (hunk 2 above)
Comment (dessalines) [reply ui/src/components/user.tsx:586]: If you want you can put it in `utils.ts`, but its fine to just put right there, as this form would get re-used anywhere that check needs to be done anyway.

**Comment by Axighi** [comment]
>I'm also still torn on how long this bio should be. Any suggestions on whether it should be a page type thing, or just a simple one-liner?

Maybe we could limit it to be 300? Say average sentence length is roughly 100 long(I googled it). And by limiting it to 300 we prefer users present themselves within 3 sentences.

**Comment by dessalines** [comment]
Sounds good to me for now. We can always change later.

File: ui/src/components/user.tsx [hunk 4]
```diff
@@ -375,6 +376,7 @@ export class User extends Component<any, UserState> {
                 )}
               </ul>
             </h5>
+            <div>{user.bio}</div>
```
Comment (dessalines) [inline ui/src/components/user.tsx:379]: Make sure this is markdown rendered, check the example in `comment-node` or `post-listing` for this.

**Comment by dessalines** [comment]
Could you merge from main again? The diff isn't showing correctly updated things.

File: ui/src/components/markdown-textarea.tsx [hunk 5]
```diff
@@ -121,7 +122,7 @@ export class MarkdownTextArea extends Component<
               required
               disabled={this.props.disabled}
               rows={2}
-              maxLength={10000}
+              maxLength={this.props.maxLength || 1000}
```
Comment (dessalines) [inline ui/src/components/markdown-textarea.tsx:125]: Do || `10000`, many comments and sidebars are going to be a lot longer than 1000 chars.

**Comment by dessalines** [comment]
One more thing, make sure the bio gets set like this too:

```
    } else if (res.op == UserOperation.SaveUserSettings) {
      const data = res.data as LoginResponse;
      UserService.Instance.login(data);
      this.state.user.bio = this.state.userSettingsForm.bio;
      this.state.userSettingsLoading = false;
      this.setState(this.state);

      window.scrollTo(0, 0);
    } else if (res.op == UserOperation.DeleteAccount) {
```

I noticed on testing the bio at the top wasn't properly updating from the settings form.

**Comment by dessalines** [comment]
Looks good, thanks!

**Pull request merged**